pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1514,6 +1515,16 @@ func (og *operationGenerator) GenerateVerifyControllerAttachedVolumeFunc(
 		return volumetypes.GeneratedOperations{}, volumeToMount.GenerateErrorDetailed("VerifyControllerAttachedVolume.FindPluginBySpec failed", err)
 	}
 
+	if volumeToMount.PluginIsAttachable {
+		cachedAttachedVolumes, _ := og.volumePluginMgr.Host.GetAttachedVolumes()
+		if cachedAttachedVolumes != nil {
+			_, volumeFound := cachedAttachedVolumes[volumeToMount.VolumeName]
+			if !volumeFound {
+				return volumetypes.GeneratedOperations{}, NewMountPreConditionFailedError(fmt.Sprintf("volume %s is not yet in node's status", volumeToMount.VolumeName))
```

Comment:
what's the difference of checking !volumeToMount.ReportedInUse?

Reply:
reportedInUse information is immediately available from DSOW but checking for attached volumes requires node informer/lister. I found kubelet's `VolumeHost` interface to be most convenient place for getting this information and besides there are already  methods that query node object via `VolumeHost` interface, so having one more method should work okay.

Reply:
since you already add the check in reconciler https://github.com/kubernetes/kubernetes/pull/106853/files#diff-e9392bf9a117fa5eda2756ca13783db044a18ad4585b9b1fad776f660dc13068R206
volumeToMount should already in node status volumeInUse, do we need to add check here?

Reply:
okay, so if I understand this correctly, you are saying we already checked `ReportedInUse` in reconciler, why do we need to check again in operation_exectutor? hmm - that is a valid point, I left it there for historic reasons, it was there and I did not wanted to move it unnecessarily.

Reply:
the check is still in line 1546? https://github.com/kubernetes/kubernetes/pull/106853/files#diff-3e8f8a6e5eca57db8e84bcc6c896f0a3b10289fe55d8e999b4bdf1cfdc0de394R1546

I am ok to not remote line 1546 check, but the check here seems newly added?

Reply:
The check for volumeAttached is newly added and it has been added to avoid exponential backoff that happens when volume is not added to node's `VolumeAttached` list.  This PR uses node object from informer as a first pass to verify if volume has been added to node's `VolumeAttached` list, inside the function `GenerateVerifyControllerAttachedVolumeFunc`, we still talk to api-server and fetch node object and verify again.

Also `ReportedInUse` check was always here - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1535 I generally do not want to remove it.

Reply:
by adding the logic in reconciler to check ReportedInUse, it already avoid exponential backoff happen. Do we need to add this additional logic here too?

Reply:
The code in reconciler *only* avoid exponential backoff for `ReportedInUse` status. We *still* need additional logic here to avoid exponential backoff if node does not have volume added to `node.Status.VolumeAttached`.

Reply:
oh, got it, I thought you only want to handle ReportedInUse case.

For VolumeAttached case, I think we kind of want to back off intentionally. If we add the logic here, reconciler will keep generating go routine constantly, will it have any overhead concern in case of large number of volumes case?

Reply:
actually it will not generate go routine here.